monix-eval/shared/src/main/scala/monix/eval/internal/TaskGatherN.scala

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2014-2019 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.eval.internal
+
+import cats.effect.concurrent.Deferred
+import monix.catnap.ConcurrentQueue
+import monix.eval.Task
+
+private[eval] object TaskGatherN {
+
+  def apply[A](
+    parallelism: Int,
+    in: Iterable[Task[A]]
+  ): Task[List[A]] = {
+    val itemSize = in.size
+
+    if (itemSize == 0) {
+      Task.pure(List.empty)
+    } else if (itemSize == 1) {
+      in.head.map(List(_))
+    } else {
+      for {
+        error <- Deferred[Task, Option[Throwable]]
+        queue <- ConcurrentQueue.bounded[Task, (Deferred[Task, A], Task[A])](itemSize)
+        pairs <- Task.traverse(in.toList)(task => Deferred[Task, A].map(p => (p, task)))
+        _     <- queue.offerMany(pairs)
+        fiber <- Task.gather(List.fill(parallelism.min(itemSize)) {
+          queue.poll.flatMap {
+            case (p, task) =>
+              task.redeemWith(
+                err => error.complete(Some(err)).attempt >> Task.raiseError(err),
+                p.complete
+              )
+          }.loopForever.start
+        })
```

Comment:
I think if we cancel entire `Task.gatherN` at this point, it can get dirty

If we turn `AlwaysAsyncExecution` and `enableAutoCancelableRunLoops` then any `flatMap` could be canceled

I think `bracketCase` could solve it. `acquire` would be this `Task.gather`, `use` as `Task.race` and `release` instead of `guarantee`. I will try to think about good test for it later